platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/DefaultIvyComponentGraphResolveState.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.external.model.ivy;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.internal.attributes.AttributeDesugaring;
+import org.gradle.internal.component.external.model.ModuleComponentResolveMetadata;
+import org.gradle.internal.component.model.ComponentArtifactMetadata;
+import org.gradle.internal.component.model.ComponentIdGenerator;
+import org.gradle.internal.component.model.ConfigurationMetadata;
+import org.gradle.internal.component.model.DefaultComponentGraphResolveState;
+
+import javax.annotation.Nullable;
+
+/**
+ * Default implementation of {@link IvyComponentGraphResolveState}.
+ */
+public class DefaultIvyComponentGraphResolveState extends DefaultComponentGraphResolveState<IvyModuleResolveMetadata, IvyModuleResolveMetadata> implements IvyComponentGraphResolveState {
+
+    public DefaultIvyComponentGraphResolveState(long instanceId, IvyModuleResolveMetadata metadata, AttributeDesugaring attributeDesugaring, ComponentIdGenerator idGenerator) {
+        super(instanceId, metadata, metadata, attributeDesugaring, idGenerator);
+    }
+
+    @Override
+    public ModuleComponentIdentifier getId() {
+        return getMetadata().getId();
+    }
+
+    @Override
+    public ModuleComponentResolveMetadata getModuleResolveMetadata() {
+        return getMetadata();
+    }
+
+    @Override
+    public IvyComponentArtifactResolveMetadata getResolveMetadata() {
+        return new DefaultIvyComponentArtifactResolveMetadata(getArtifactMetadata());
+    }
```

Comment:
💭 This bit is quite confusing.  I don't have any concrete suggestions to improve it, just a lot of different kinds of metadata in play.

Reply:
Yeah I 100% agree there are way too many interfaces involved here. This whole part of the code is in a middle-ground state right now and this change hopes to be a small part of that. 

In practice, it makes sense to have a different metadata/state type for ivy components compared to other components, because ivy components have configurations -- while other components do not (local Gradle components do have configurations for now, but this PR is the start of the work to remove the concept of configurations from local components from a DM perspective).